core/pta/rtc.c

```diff
@@ -0,0 +1,180 @@
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+	struct optee_rtc_time time = { };
+	struct pta_rtc_time *pta_time = NULL;
+
+	if (types != TEE_PARAM_TYPES(TEE_PARAM_TYPE_MEMREF_OUTPUT,
+				     TEE_PARAM_TYPE_NONE,
+				     TEE_PARAM_TYPE_NONE,
+				     TEE_PARAM_TYPE_NONE)) {
+		return TEE_ERROR_BAD_PARAMETERS;
+	}
+
+	pta_time = params[0].memref.buffer;
+	if (!pta_time || params[0].memref.size != sizeof(*pta_time))
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	res = rtc_get_time(&time);
+	if (res)
+		return res;
+
+	rtc_pta_copy_time_from_optee(pta_time, &time);
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result rtc_pta_set_time(uint32_t types,
+				   TEE_Param params[TEE_NUM_PARAMS])
+{
+	struct optee_rtc_time time = { };
+	struct pta_rtc_time *pta_time = NULL;
+
+	if (types != TEE_PARAM_TYPES(TEE_PARAM_TYPE_MEMREF_INPUT,
+				     TEE_PARAM_TYPE_NONE,
+				     TEE_PARAM_TYPE_NONE,
+				     TEE_PARAM_TYPE_NONE)) {
+		return TEE_ERROR_BAD_PARAMETERS;
+	}
+
+	pta_time = params[0].memref.buffer;
+	if (!pta_time || params[0].memref.size != sizeof(*pta_time))
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	time.tm_sec = pta_time->tm_sec;
+	time.tm_min = pta_time->tm_min;
+	time.tm_hour = pta_time->tm_hour;
+	time.tm_mday = pta_time->tm_mday;
+	time.tm_mon = pta_time->tm_mon;
+	time.tm_year = pta_time->tm_year;
+	time.tm_wday = pta_time->tm_wday;
+
+	return rtc_set_time(&time);
+}
+
+static TEE_Result rtc_pta_set_offset(uint32_t types,
+				     TEE_Param params[TEE_NUM_PARAMS])
+{
+	if (types != TEE_PARAM_TYPES(TEE_PARAM_TYPE_VALUE_INPUT,
+				     TEE_PARAM_TYPE_NONE,
+				     TEE_PARAM_TYPE_NONE,
+				     TEE_PARAM_TYPE_NONE)) {
+		return TEE_ERROR_BAD_PARAMETERS;
+	}
+
+	rtc_set_offset((int32_t)params[0].value.a);
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result rtc_pta_get_offset(uint32_t types,
+				     TEE_Param params[TEE_NUM_PARAMS])
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+	long offset = 0;
+
+	if (types != TEE_PARAM_TYPES(TEE_PARAM_TYPE_VALUE_OUTPUT,
+				     TEE_PARAM_TYPE_NONE,
+				     TEE_PARAM_TYPE_NONE,
+				     TEE_PARAM_TYPE_NONE)) {
+		return TEE_ERROR_BAD_PARAMETERS;
+	}
+
+	res = rtc_get_offset(&offset);
+	if (res)
+		return res;
+
+	params[0].value.a = (uint32_t)offset;
+
+	return res;
+}
+
+static TEE_Result rtc_pta_get_info(uint32_t types,
+				   TEE_Param params[TEE_NUM_PARAMS])
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+	struct pta_rtc_info *info = NULL;
+	struct optee_rtc_time range_min = { };
+	struct optee_rtc_time range_max = { };
+	uint64_t features = 0;
+
+	if (types != TEE_PARAM_TYPES(TEE_PARAM_TYPE_MEMREF_OUTPUT,
+				     TEE_PARAM_TYPE_NONE,
+				     TEE_PARAM_TYPE_NONE,
+				     TEE_PARAM_TYPE_NONE)) {
+		return TEE_ERROR_BAD_PARAMETERS;
+	}
+
+	info = params[0].memref.buffer;
+	if (!info || params[0].memref.size != sizeof(*info))
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	res = rtc_get_info(&features, &range_min, &range_max);
+	if (res)
+		return res;
+
+	info->version = PTA_RTC_INFO_VERSION;
```

Comment:
can you add `memset(info, 0, sizeof(*info));` before filling with data?

Reply:
Yes